File: evals/parity/reference/sc.R
# Synthetic-control parity reference (R).
# Canonical classic R SC: Synth (Abadie, Diamond, Hainmueller) — simplex-constrained
# weights (>= 0, sum to 1). `df` is preloaded by the parity runner from
# fixtures/sc_parity.csv.
#
# Prints, for cross-language comparison:
#   ATT       : mean post-treatment gap (actual treated - synthetic)
#   RMSPE_PRE : pre-treatment root mean squared prediction error (fit quality)
#   W_DONOR1  : synthetic-control weight on donor_1
suppressMessages(library(Synth))

TREAT_TIME <- 13
TREATED <- "treated"

# Synth needs a numeric unit id and a unit-name column.
df$unit <- as.character(df$unit)
units <- sort(unique(df$unit))
df$unit_num <- match(df$unit, units)
treated_num <- match(TREATED, units)
donor_nums <- setdiff(seq_along(units), treated_num)

time_min <- min(df$time)
pre_times <- time_min:(TREAT_TIME - 1)
post_times <- TREAT_TIME:max(df$time)

# Match on every pre-treatment outcome (the full pre-period path), mirroring the
# Python recipe's reliance on the pre-treatment outcome trajectory.
special_preds <- lapply(pre_times, function(tt) list("outcome", tt, "mean"))

dp <- dataprep(
  foo = as.data.frame(df),
  predictors = NULL,
  dependent = "outcome",
  unit.variable = "unit_num",
  time.variable = "time",
  special.predictors = special_preds,
  treatment.identifier = treated_num,
  controls.identifier = donor_nums,
  time.predictors.prior = pre_times,
  time.optimize.ssr = pre_times,
  unit.names.variable = "unit",
  time.plot = time_min:max(df$time)
)

# Default Synth: simplex-constrained weights via nested V-optimization.
so <- synth(dp, verbose = FALSE, optimxmethod = "BFGS")

w <- as.numeric(so$solution.w)
names(w) <- units[donor_nums]
w_donor1 <- unname(w["donor_1"])

synthetic <- as.numeric(dp$Y0plot %*% so$solution.w)
actual <- as.numeric(dp$Y1plot)
times <- as.integer(rownames(dp$Y1plot))

gap <- actual - synthetic
pre_mask <- times < TREAT_TIME
post_mask <- times >= TREAT_TIME

att <- mean(gap[post_mask])
rmspe_pre <- sqrt(mean(gap[pre_mask]^2))

cat(sprintf("ATT:%f\n", att))
cat(sprintf("RMSPE_PRE:%f\n", rmspe_pre))
cat(sprintf("W_DONOR1:%f\n", w_donor1))

File: evals/parity/reference/sc.py
# Synthetic-control parity reference (Python).
# Canonical modern Python SC: scpi_pkg (Cattaneo, Feng, Palomba, Titiunik) — the
# package the Python template already uses. Classic simplex-constrained SCM
# (Abadie). `df` is preloaded by the parity runner from fixtures/sc_parity.csv.
#
# Prints, for cross-language comparison:
#   ATT     : mean post-treatment gap (actual treated - synthetic)
#   RMSPE_PRE : pre-treatment root mean squared prediction error (fit quality)
#   W_DONOR1  : synthetic-control weight on donor_1
import numpy as np
from scpi_pkg.scdata import scdata
from scpi_pkg.scest import scest

TREAT_TIME = 13
TREATED = "treated"

time_min = int(df["time"].min())
time_max = int(df["time"].max())
donors = sorted(u for u in df["unit"].unique() if u != TREATED)

sc_data = scdata(
    df=df, id_var="unit", time_var="time", outcome_var="outcome",
    period_pre=np.arange(time_min, TREAT_TIME),
    period_post=np.arange(TREAT_TIME, time_max + 1),
    unit_tr=TREATED, unit_co=donors, constant=False, cointegrated_data=False,
)

# Default w_constr is the canonical Abadie simplex (weights >= 0, sum to 1).
est = scest(sc_data, w_constr={"name": "simplex"})

# est.w preserves the unit_co order; align weights positionally to `donors`.
# (scpi relabels "donor_1" -> "donor 1" for display, so don't match on labels.)
weights = dict(zip(donors, est.w.iloc[:, 0].to_numpy()))
w_donor1 = float(weights["donor_1"])

# Reconstruct synthetic outcome over the full horizon from donor panel x weights.
donor_panel = df[df["unit"].isin(donors)].pivot(index="time", columns="unit",
                                                values="outcome")
weight_vec = np.array([float(weights[d]) for d in donor_panel.columns])
synthetic = donor_panel.to_numpy() @ weight_vec
actual = (df[df["unit"] == TREATED].sort_values("time")["outcome"].to_numpy())
times = donor_panel.index.to_numpy()

gap = actual - synthetic
pre_mask = times < TREAT_TIME
post_mask = times >= TREAT_TIME

att = float(gap[post_mask].mean())
rmspe_pre = float(np.sqrt(np.mean(gap[pre_mask] ** 2)))

print(f"ATT:{att}")
print(f"RMSPE_PRE:{rmspe_pre}")
print(f"W_DONOR1:{w_donor1}")
